File: android/fundamental-core/src/main/kotlin/com/fundamental/core/engine/Geometry.kt
package com.fundamental.core.engine

import com.fundamental.core.math.EPS
import com.fundamental.core.math.Vec3
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

// Geometry helpers — the Kotlin port of swift/Sources/FundamentalCore/Math/Geometry.swift. `Box` itself
// lives in engine/Types.kt; this file adds the dipole geometry and box queries the integrator and the
// magnetism/charge field() hooks use.

/** Box + heading (unit Vec3) + polarity spin — the dipole axis. */
data class AxisBox(val box: Box, val heading: Vec3, val spin: Float)

/** One pole of a dipole: a 3D position and signed charge (±1). */
data class Pole(val position: Vec3, val charge: Float)

/** The nearest point of the filled box to [p]: closest boundary point outside, `p` itself inside. */
fun nearestOnBox(p: Vec3, b: Box): Vec3 {
    val lo = b.center - b.halfExtents
    val hi = b.center + b.halfExtents
    fun c(v: Float, l: Float, h: Float) = if (v < l) l else if (v > h) h else v
    return Vec3(c(p.x, lo.x, hi.x), c(p.y, lo.y, hi.y), c(p.z, lo.z, hi.z))
}

/** Signed distance to the box: negative inside, zero on the edge, positive outside (3D box SDF). */
fun sdfBox(p: Vec3, b: Box): Float {
    val q = Vec3(abs(p.x - b.center.x), abs(p.y - b.center.y), abs(p.z - b.center.z)) - b.halfExtents
    val outside = Vec3(max(q.x, 0f), max(q.y, 0f), max(q.z, 0f)).length()
    val inside = min(max(q.x, max(q.y, q.z)), 0f)
    return outside + inside
}

/**
 * The two poles of the body's dipole, laid on its heading axis at the box edge. The `+` (N) pole sits
 * at `+heading`; `spin < 0` swaps them. Reach is proportional to the box extent along the heading.
 */
fun polePair(b: AxisBox): Pair<Pole, Pole> {
    val he = b.box.halfExtents
    val h = b.heading
    val tx = if (h.x != 0f) he.x / abs(h.x) else Float.POSITIVE_INFINITY
    val ty = if (h.y != 0f) he.y / abs(h.y) else Float.POSITIVE_INFINITY
    val tz = if (h.z != 0f) he.z / abs(h.z) else Float.POSITIVE_INFINITY
    val reach = min(tx, min(ty, tz))
    val axis = h * reach
    val s = if (b.spin < 0f) -1f else 1f
    return Pair(
        Pole(b.box.center + axis, s),
        Pole(b.box.center - axis, -s),
    )
}

/** Superposition of a set of poles' radial q/d² fields at [p] — the bar-magnet / dipole field. */
fun dipoleField(poles: List<Pole>, p: Vec3): Vec3 {
    var f = Vec3.ZERO
    for (pole in poles) {
        val d = p - pole.position
        val dn = max(EPS, d.length())
        val k = pole.charge / (dn * dn)
        f += (d / dn) * k
    }
    return f
}

/**
 * The net *structure* field at a point: the superposition of every visible body's `field()`
 * contribution, with the same range cull as the integrator. The field streamlines draw and `fieldflow`
 * follows. Pure — no env mutation, safe mid-integration. (Swift `netField`, streamlines.swift.)
 */
fun netField(bodies: List<Body>, forces: ForceRegistry, point: Vec3): Vec3 {
    var f = Vec3.ZERO
    for (b in bodies) {
        if (!b.isVisible || b.tokens.isEmpty()) continue
        if (b.range > 0f) {
            val delta = b.center - point
            if (delta.lengthSquared() >= b.range * b.range * 2.56f) continue
        }
        for (tok in b.tokens) {
            forces[tok]?.field(b, point)?.let { f += it }
        }
    }
    return f
}

File: rust/crates/fundamental-core/src/math/geometry.rs
//! Field geometry — the renderable/followable STRUCTURE a body radiates (#1041).
//!
//! Mirrors `packages/core/src/math/geometry.ts`. These are the formulas behind
//! [`Force::field`](crate::engine::Force::field): the same superposition a field-line diagram traces
//! and `fieldflow` follows, so the picture is the engine's real field rather than a hand-rolled
//! stand-in.
//!
//! A structure field is **not** a particle path. `apply()` is unchanged by any of this — a sideways
//! moving particle can orbit a gravity well rather than fall down its field line.

/// Floors the distance at a pole/core so a sample never divides by zero. JS `geometry.EPS`.
pub const EPS: f64 = 1.0;

/// A point source: position and signed strength.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Pole {
    pub x: f64,
    pub y: f64,
    pub q: f64,
}

/// The two poles of a body's dipole, laid on its heading axis at the box edge.
///
/// The `+` (N) pole sits at the `+(ux, uy)` end and the `−` (S) pole at the other; `spin < 0` swaps
/// them. `reach` is the centre-to-edge distance along the heading, so a wide element makes a long
/// magnet and a tall one a short, fat one. Polarity and geometry only — strength is the consumer's.
pub fn pole_pair(cx: f64, cy: f64, ux: f64, uy: f64, hw: f64, hh: f64, spin: f64) -> [Pole; 2] {
    let tx = if ux != 0.0 { hw / ux.abs() } else { f64::INFINITY };
    let ty = if uy != 0.0 { hh / uy.abs() } else { f64::INFINITY };
    let reach = tx.min(ty); // centre → box edge along the heading
    let (ax, ay) = (ux * reach, uy * reach);
    let s = if spin < 0.0 { -1.0 } else { 1.0 };
    [
        Pole { x: cx + ax, y: cy + ay, q: s },
        Pole { x: cx - ax, y: cy - ay, q: -s },
    ]
}

/// The in-plane field of a set of poles: the superposition of each pole's radial `q/d²` monopole
/// (away from `+`, toward `−`). For a body's two poles this is the classic dipole, whose streamlines
/// are the bar-magnet (N→S) diagram.
pub fn dipole_field(poles: &[Pole], x: f64, y: f64) -> (f64, f64) {
    let (mut fx, mut fy) = (0.0, 0.0);
    for p in poles {
        let dx = x - p.x;
        let dy = y - p.y;
        let d = (dx * dx + dy * dy).sqrt().max(EPS);
        let k = p.q / (d * d); // radial monopole, 1/d² falloff
        fx += (dx / d) * k;
        fy += (dy / d) * k;
    }
    (fx, fy)
}

/// The radial monopole field of a single point charge: straight lines OUT of a `+` source
/// (`sign >= 0`) and IN to a `−`. `E = sgn·s·r̂/d²`.
pub fn monopole_field(cx: f64, cy: f64, sign: f64, s: f64, x: f64, y: f64) -> (f64, f64) {
    let dx = x - cx;
    let dy = y - cy;
    let d = (dx * dx + dy * dy).sqrt().max(EPS);
    let sgn = if sign < 0.0 { -1.0 } else { 1.0 };
    let mag = (sgn * s) / (d * d); // 1/d², signed by polarity
    ((dx / d) * mag, (dy / d) * mag)
}

/// The inward radial gravitational field: `g = M·r̂/d²`, always TOWARD the mass.
pub fn gravity_field(cx: f64, cy: f64, mass: f64, x: f64, y: f64) -> (f64, f64) {
    let dx = cx - x; // toward the body — gravity attracts
    let dy = cy - y;
    let d = (dx * dx + dy * dy).sqrt().max(EPS);
    let mag = mass / (d * d);
    ((dx / d) * mag, (dy / d) * mag)
}
